# taken from https://hbctraining.github.io/scRNA-seq/lessons/cell_cycle_scoring.html
# and genes from https://github.com/hbc/tinyatlas/blob/master/cell_cycle/Mus_musculus.csv

suppressPackageStartupMessages({
  library(plyr)
  library(dplyr)
  library(Seurat)
  library(RCurl)
  library(AnnotationHub)
})

script.dir <- dirname(sys.frame(1)$ofile)
source(paste0(script.dir, "/seurat_RNA.R"))

# TODO: while the current approach works for RNA-seq, may need to take a differing approach with ATAC-seq to match ENSDB loci to peaks instead of by gene

## quickCycleTest
# generates a DimPlot for a set of genes linked to cell cycle for quick identification of cell populations undergoing mitosis
# suob: seurat object containing the relevant genes as activity features
quickCycleTest <- function(suob) {
  X11()
  p = FeaturePlot(suob, features = c("Pcna", "Top2a", "Mcm6", "Mki67"), ncol=3)
  print(p)
}

## cellCycleAnnotate
# annotate cells with cell cycle stage information based on genes known to be highly associated with each stage of the cell cycle in Mus Musculus, with cycle stage annotations placed in the $Phase parameter
# suob: seurat object with activity levels in each of the relevant genes
cellCycleAnnotate <- function(suob, plot=TRUE) {
  # for Mus Musculus
  cc_file <- getURL("https://raw.githubusercontent.com/hbc/tinyatlas/master/cell_cycle/Mus_musculus.csv")
  cell_cycle_genes <- read.csv(text = cc_file)
  ah <- AnnotationHub()
  ahDb <- query(ah, pattern = c("Mus musculus", "EnsDb"), ignore.case = TRUE)
  id <- ahDb %>% mcols() %>% rownames() %>% tail(n = 1)
  edb <- ah[[id]]
  annotations <- genes(edb, return.type = "data.frame")
  annotations <- annotations %>% dplyr::select(gene_id, gene_name, seq_name, gene_biotype, description)
  cell_cycle_markers <- dplyr::left_join(cell_cycle_genes, annotations, by = c("geneID" = "gene_id"))
  s_genes <- cell_cycle_markers %>% dplyr::filter(phase == "S") %>% pull("gene_name")
  g2m_genes <- cell_cycle_markers %>% dplyr::filter(phase == "G2/M") %>% pull("gene_name")
  suob <- CellCycleScoring(suob, g2m.features = g2m_genes, s.features = s_genes)
  if (plot) {
    X11()
    p = DimPlot(suob, group.by = 'Phase')
    print(p)
  }
  return(suob)
}

## regressCellCycle
# use cell cycle scores determined in cellCycleAnnotate to regress out factors associated with cell division so that these effects do not contribute to the clustering of expression level heterogeneity
# suob: seurat object with S.Score and G2M.Score attributes (as generated by cellCycleAnnotate)
# standard_filters: use standard seurat filters for PCA and clustering, or turn this off to enter them manually
# verbose: print extra figures
regressCellCycle <- function(suob, standard_filters=TRUE, verbose=FALSE, plot=TRUE) {
  X11()
  p = RidgePlot(suob, features = c("Pcna", "Top2a", "Mcm6", "Mki67"), ncol = 2, group.by = 'Phase')
  print(p)
  variablefeatures <- VariableFeatures(suob)
  suob <- ScaleData(suob, vars.to.regress = c("S.Score", "G2M.Score"), features = variablefeatures)
  num_dims <- 50
  suob <- RunPCA(suob, features = VariableFeatures(object = suob), npcs = strtoi(num_dims, base = 0L))
  print(suob[["pca"]], dims = 1:5, nfeatures = 5)
  if(verbose) {
    X11()
    p = VizDimLoadings(suob, dims = 1:2, reduction = "pca")
    print(p)
    X11()
    p = DimPlot(suob, reduction = "pca", group.by = "Phase")
    print(p)
    X11()
    p = DimHeatmap(suob, dims = 1:15, cells = 500, balanced = TRUE, fast=FALSE)
    print(p)
  }
  X11()
  p = ElbowPlot(suob, ndims = 50)
  print(p)
  if(standard_filters) {
    clu_dims <- 50
  } else {
    clu_dims <- typeline("Clustering: Enter number of PCA dims (consider elbow plot): ")
    print(strtoi(clu_dims))
  }
  res <- 0.8
  suob <- FindNeighbors(suob, dims = 1:strtoi(clu_dims, base = 0L))
  suob <- FindClusters(suob, resolution = as.double(res))
  head(Idents(suob), 5)
  suob <- RunUMAP(suob, dims = 1:strtoi(clu_dims))
  if (plot) {
    X11()
    p = DimPlot(suob, reduction="umap")
    print(p)
  }
  return(suob)
}