suppressPackageStartupMessages({
  library(plyr)
  library(dplyr)
  library(Seurat)
  library(patchwork)
  library(Matrix)
  library(hdf5r)
  library(Signac)
  library(EnsDb.Mmusculus.v79)
})

### TODO ###
# - review mutation naming scheme, may wish to switch it from X1-110054321 to something like chr1-110054321 which makes much more sense, and also consider making it back to 1 indexing for better compatibility with VCF files later on

### Helper Functions ###

## indexMutMeta
# add an index field that contains mutation names which are compatible with those typically generated by ReadMut
indexMutMeta <- function(suob.mut.meta) {
  suob.mut.meta$index <- paste0("X", suob.mut.meta$chromosome, "-", (suob.mut.meta$position-1)) # the subtraction is only done because vartrix operates in 0-indexing whereas the original vcfs are in 1-indexing
  return(suob.mut.meta)
}

indexMutMetaUnder <- function(suob.mut.meta) {
  suob.mut.meta$index <- paste0("X", suob.mut.meta$chromosome, "_", (suob.mut.meta$position-1)) # the subtraction is only done because vartrix operates in 0-indexing whereas the original vcfs are in 1-indexing
  return(suob.mut.meta)
}

### Major Functionality ###

## ReadMut
# Reads a mutation matrix generated by run_vartrix.sh into an object of ALT and REF allele counts
# path.to.directory: the location of the directory output by vartrix containing alt_matrix.mtx, ref_matrix.mtx, variants.tsv, and barcodes.tsv 
ReadMut <- function(path.to.directory) {
  path_to_alt_variants <- paste(path.to.directory, "/alt_matrix.mtx", sep="")
  path_to_ref_variants <- paste(path.to.directory, "/ref_matrix.mtx", sep="")
  path_to_variant_list <- paste(path.to.directory, "/variants.tsv", sep="")
  path_to_cell_list <- paste(path.to.directory, "/barcodes.tsv", sep="")
  mut.varnames <- readLines(con = path_to_variant_list)
  mut.varnames <- str_remove(mut.varnames, "chr")
  mut.varnames <- paste0("X", mut.varnames)
  mut.varnames <- make.names(mut.varnames, unique=TRUE)
  mut.cellnames <- readLines(con = path_to_cell_list)
  mut.alt <- readMM(file = path_to_alt_variants)
  rownames(x = mut.alt) <- mut.varnames
  colnames(x = mut.alt) <- mut.cellnames
  mut.ref <- readMM(file = path_to_ref_variants)
  rownames(x = mut.ref) <- mut.varnames
  colnames(x = mut.ref) <- mut.cellnames
  mut <- list()
  mut[["ALT"]] <- mut.alt
  mut[["REF"]] <- mut.ref
  return(mut)
}

## loadSeuratWithMutations
# loads the REF-allele and ALT-allele matrices into the seurat object as assays "REF" and "ALT" respectively, preserving the "RNA" assay, and creates two new assays "ALTB" and "REFB" which are binary representations of whether the allele is present in a cell in any capacity
# note: this function makes sure that the cells within the RNA seurat object match those within the mutation matrices, and only keeps those which match (if no cell filtering was done during the mutation counting step, then the agreement between the two should be unanimous)
# suob.rna: a suob object of an RNA assay (e.g. generated by ReadRNA)
# suob.mut: a mutation object generated by ReadMut
# projectname: the name of the project to be used
# output: a seurat object with assays "RNA", "REF", "ALT", "REFB", and "ALTB"
loadSeuratWithMutations <- function(suob.rna, suob.mut, projectname, atac=FALSE) {
  suob.rna.matched <- suob.rna[,(colnames(suob.rna) %in% colnames(suob.mut[["ALT"]]))]
  suob.mut.alt.matched <- suob.mut[["ALT"]][,(colnames(suob.mut[["ALT"]]) %in% colnames(suob.rna))]
  suob.mut.ref.matched <- suob.mut[["REF"]][,(colnames(suob.mut[["REF"]]) %in% colnames(suob.rna))]
  cat("Expression Cells before filtering: ", ncol(suob.rna), "\n")
  cat("Mutation Cells before filtering: ", ncol(suob.mut[["ALT"]]), "\n")
  cat("Expression Cells before filtering: ", ncol(suob.rna.matched), "\n")
  cat("Mutation Cells before filtering: ", ncol(suob.mut.alt.matched), "\n")
  if (atac) {
    assay <- CreateChromatinAssay(counts = suob.rna.matched, sep = c(":", "-"), project = projectname)
    suob <- CreateSeuratObject(counts = assay, assay = "peaks")
  } else {
    suob <- CreateSeuratObject(counts = suob.rna.matched, project = projectname)
  }
  suob[["ALT"]] <- CreateAssayObject(counts = suob.mut.alt.matched)
  suob[["REF"]] <- CreateAssayObject(counts = suob.mut.ref.matched)
  suob[["ALTB"]] <- CreateAssayObject(counts = as.data.frame(suob.mut.alt.matched > 0))
  suob[["REFB"]] <- CreateAssayObject(counts = as.data.frame(suob.mut.ref.matched > 0))
  suob.mut.vaf <- suob.mut.alt.matched/(suob.mut.ref.matched+suob.mut.alt.matched)
  suob.mut.vaf[is.na(suob.mut.vaf)] <- 0
  suob[["VAF"]] <- CreateAssayObject(counts = suob.mut.vaf)
  return(suob)
}

## loadSeuratATACWithMutations
# creates the seurat ATAC object and loads the REF-allele and ALT-allele matrices into that object as assays "REF" and "ALT" respectively, preserving the "ATAC" assay, and creates two new assays "ALTB" and "REFB" which are binary representations of whether the allele is present in a cell in any capacity
# note: replaces ReadATAC due to the need to filter the metadata file after checking to make sure that the cells within the RNA seurat object match those within the mutation matrices, only keeping those which match (if no cell filtering was done during the mutation counting step, then the agreement between the two should be unanimous)
# path.to.peaks: path to peaks output of 10X cellranger (usually filtered_peak_bc_matrix.h5)
# path.to.metadata: path to metadata of cells output from cellranger (usually singlecell.csv)
# path.to.fragments: path to compressed fragments file output from cellranger (usually fragments.tsv.gz)
# suob.mut: a mutation object generated by ReadMut
# projectname: the name of the project to be used
# output: a seurat object with assays "ATAC", "REF", "ALT", "REFB", and "ALTB"
loadSeuratATACWithMutations <- function(path.to.peaks, path.to.metadata, path.to.fragments, suob.mut, projectname) {
  counts <- Read10X_h5(filename = path.to.peaks)
  metadata <- read.csv(file = path.to.metadata, header = TRUE, row.names = 1)
  counts.matched <- counts[,(colnames(counts) %in% colnames(suob.mut[["ALT"]]))]
  metadata.matched <- metadata[(rownames(metadata) %in% colnames(suob.mut[["ALT"]])),]
  suob.mut.alt.matched <- suob.mut[["ALT"]][,(colnames(suob.mut[["ALT"]]) %in% colnames(counts))]
  suob.mut.ref.matched <- suob.mut[["REF"]][,(colnames(suob.mut[["REF"]]) %in% colnames(counts))]
  cat("Peak Cells before filtering: ", ncol(counts), "\n")
  cat("Mutation Cells before filtering: ", ncol(suob.mut[["ALT"]]), "\n")
  cat("Peak Cells after filtering: ", ncol(counts.matched), "\n")
  cat("Mutation Cells after filtering: ", ncol(suob.mut.alt.matched), "\n")
  chrom_assay <- CreateChromatinAssay(counts = counts.matched, sep = c(":", "-"), genome = 'mm10', fragments = path.to.fragments)
  suob <- CreateSeuratObject(counts = chrom_assay, assay = "PEAK", meta.data = metadata.matched)
  annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)
  seqlevelsStyle(annotations) <- 'UCSC'
  genome(annotations) <- "mm10"
  Annotation(suob) <- annotations
  suob[["ALT"]] <- CreateAssayObject(counts = suob.mut.alt.matched)
  suob[["REF"]] <- CreateAssayObject(counts = suob.mut.ref.matched)
  suob[["ALTB"]] <- CreateAssayObject(counts = as.data.frame(suob.mut.alt.matched > 0))
  suob[["REFB"]] <- CreateAssayObject(counts = as.data.frame(suob.mut.ref.matched > 0))
  suob.mut.vaf <- suob.mut.alt.matched/(suob.mut.ref.matched+suob.mut.alt.matched)
  suob.mut.vaf[is.na(suob.mut.vaf)] <- 0
  suob[["VAF"]] <- CreateAssayObject(counts = suob.mut.vaf)
  return(suob)
}
